### 1. Implement Naïve Bayes method using scikit-learn library
Use dataset available with name glass
Use train_test_split to create training and testing part
Evaluate the model on test part using score and 

In [1]:
#Importing packages to create arrays
import pandas as pd
import numpy as np

In [2]:
#loading data set
df=pd.read_csv("glass.csv")

In [3]:
#to view the structure of data
df.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [4]:
df.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [6]:
#removing the column to be predicted from the dataframe 'df'
pred_df=df.drop(["Type"],axis=1)

#creating an array for the target column (type) for prediction 
y=df["Type"]

In [7]:
#splitting data(both X,y) to train the model & test

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(pred_df, y, test_size= 0.8 ,random_state=0)

In [8]:
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics
from sklearn.metrics import accuracy_score

In [9]:
#1) Implement Naïve Bayes method using scikit-learn library

gnb = GaussianNB()

##Training data with Gaussian Naive Bayes
# Training on X_Train & y_Train
#Applying prediction on X_Test
y_pred = gnb.fit(X_train, y_train).predict(X_test)
# calculating the accuracy score(accuracy score is calculated on predicted value & Actual value)
acc_nb=accuracy_score(y_test,y_pred)
print("Naïve Bayes Accuracy is:", acc_nb)

Naïve Bayes Accuracy is: 0.5348837209302325


In [10]:
#Calculating % and rounding off to 2 digits
print("Naïve Bayes Accuracy is:")
round(acc_nb * 100,2)

Naïve Bayes Accuracy is:


53.49

In [11]:
## Classification report is used to measure the quality of predictions using classification algorithm and accuracy
print(metrics.classification_report(y_test,y_pred, zero_division=0))

              precision    recall  f1-score   support

           1       0.58      0.23      0.33        60
           2       0.45      0.72      0.55        61
           3       0.00      0.00      0.00        15
           5       0.47      0.88      0.61         8
           6       0.55      1.00      0.71         6
           7       0.88      0.95      0.91        22

    accuracy                           0.53       172
   macro avg       0.49      0.63      0.52       172
weighted avg       0.52      0.53      0.48       172



### 2. Implement linear SVM method using scikit library
Use the same dataset above
Use train_test_split to create training and testing part
Evaluate the model on test part using score and 

In [12]:
from sklearn.svm import SVC, LinearSVC

#Training data with Support vector Classification
svc = SVC(kernel ='linear', C = 5)
# Training on X_Train & y_Train
svc.fit(X_train, y_train)
#Applying prediction on X_Test
y_pred = svc.predict(X_test)
# calculating the accuracy score(accuracy score is calculated on predicted value & Actual value)
acc_svc=accuracy_score(y_test,y_pred)
print("LinearSVM Accuracy is:", acc_svc)

LinearSVM Accuracy is: 0.622093023255814


In [13]:
#Calculating % and rounding off to 2 digits
print("SVM Accuracy is:")
round(acc_svc * 100 , 2)

SVM Accuracy is:


62.21

In [15]:
## Classification report is used to measure the quality of predictions using classification algorithm and accuracy
print(metrics.classification_report(y_test,y_pred, zero_division=0))

              precision    recall  f1-score   support

           1       0.67      0.62      0.64        60
           2       0.54      0.67      0.60        61
           3       0.00      0.00      0.00        15
           5       0.47      0.88      0.61         8
           6       0.67      0.67      0.67         6
           7       0.95      0.82      0.88        22

    accuracy                           0.62       172
   macro avg       0.55      0.61      0.57       172
weighted avg       0.59      0.62      0.60       172



### Which algorithm you got better accuracy? Can you justify why?

* For a classification task, achieving an accuracy score of 100% is considered ideal. 
* In the case of applying the Gaussian Naive Bayes algorithm to a particular dataset, the accuracy obtained was 53.49%. On the other hand, the Support Vector Machine (SVM) with a linear kernel achieved a higher accuracy score of 62.21%. 

* This indicates that, for this specific dataset, the Linear SVM model outperforms the Gaussian Naive Bayes model, as evidenced by its superior accuracy score.

* If the data set has many features and is easy to interpret, Naive Bayes can be more effective.
The data distribution, Naive Bayes assumes that the features are independent, if the features in the data are not independent, the performance of Naive Bayes can decrease.

The accuracy of the model can also depend on the choice of hyperparameters, if the hyperparameters of the Naïve Bayes algorithm were not optimized, this could result in a lower accuracy score. It is important to keep in mind that accuracy is not the only evaluation metric that should be considered, other metrics such as precision, recall, and F1-score should also be considered to get a comprehensive evaluation of the model's performance.